common/src/VisClient.cpp

```diff
@@ -0,0 +1,825 @@
+        mWs->send(buffer);
+        message.assign(buffer);
+        return mRequestId;
+    } else {
+        ALOGE("Unable to send command through, no websocket.");
+    }
+    return -1;
+}
+
+Status VisClient::doSubscribeProperty(const std::string& propertyName, CommandHandler observer,
+                                      std::future<WMessageResult>& future) {
+    if (mConnectedState == ConnState::STATE_DISCONNECTED) {
+        return Status::INVALID_STATE;
+    }
+
+    std::string message;
+    int id = sendWSMessage(requestTemplateSubscribe, propertyName.c_str(), message);
+
+    if (id < 0) {
+        return Status::UNKNOWN_ERROR;
+    }
+
+    SubscriptionRequest sr{observer, propertyName, static_cast<size_t>(id)};
+
+    mSubscriptionObservers.insert(std::make_pair(id, sr));
+    std::promise<WMessageResult> p;
+    future = p.get_future();
+    mPromised.emplace(std::pair<int, std::promise<WMessageResult> >(id, std::move(p)));
+    return Status::OK;
+}
+
+Status VisClient::subscribeProperty(const std::string& propertyName, CommandHandler observer,
+                                    std::future<WMessageResult>& future) {
+    std::lock_guard<std::mutex> lock(mLock);
+    return doSubscribeProperty(propertyName, observer, future);
+}
+
+Status VisClient::unsubscribe(const std::string& subscriptionId, CommandHandler observer,
+                              std::future<WMessageResult>& sfuture) {
+    std::lock_guard<std::mutex> lock(mLock);
+    if (mConnectedState == ConnState::STATE_DISCONNECTED) {
+        return Status::INVALID_STATE;
+    }
+
+    int id = sendWSMessage(requestTemplateUnsubscribe, subscriptionId.c_str());
+
+    if (id < 0) {
+        return Status::UNKNOWN_ERROR;
+    }
+    SubscriptionRequest sr{observer, subscriptionId, static_cast<size_t>(id)};
+    mSubscriptionObservers.insert(std::make_pair(id, sr));
+    std::promise<WMessageResult> p;
+    sfuture = p.get_future();
+    mPromised.emplace(std::pair<int, std::promise<WMessageResult> >(id, std::move(p)));
+    return Status::OK;
+}
+
+Status VisClient::unsubscribeAll(std::future<WMessageResult>& future) {
+    std::lock_guard<std::mutex> lock(mLock);
+    if (mConnectedState == ConnState::STATE_DISCONNECTED) {
+        return Status::INVALID_STATE;
+    }
+
+    int id = sendWSMessage(requestTemplateUnsubscribeAll, "unsubscribeAll");
+
+    if (id < 0) {
+        return Status::UNKNOWN_ERROR;
+    }
+    std::promise<WMessageResult> p;
+    future = p.get_future();
+    mPromised.emplace(std::pair<int, std::promise<WMessageResult> >(id, std::move(p)));
+    return Status::OK;
+}
+
+/* uWS specific handlers*/
+
+void VisClient::handleServerDisconnect() {
+    mConnectedState = ConnState::STATE_DISCONNECTED;
+    mWs = nullptr;
+}
+
+void VisClient::handleError(void* user) {
+    ALOGV("handleError");
+    std::lock_guard<std::mutex> lock(mLock);
+
+    switch (reinterpret_cast<size_t>(user)) {
+        case 1:
```

Comment:
As I understand these error codes are defined somewhere outside of project.
If this is the case, we may provide comment about source of these numbers, for possible expansion and verification.